ocaml/xapi/xapi_vm_group_helpers.ml

```diff
@@ -0,0 +1,173 @@
+(*
+ * Copyright (C) 2024 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "xapi_vm_group_helpers" end)
+
+open D
+
+let check_breach ~__context ~group =
+  let vms =
+    Db.VM_group.get_VMs ~__context ~self:group
+    |> List.filter (fun vm ->
+           Db.VM.get_power_state ~__context ~self:vm = `Running
+       )
+  in
+  let get_host vm =
+    let host =
+      match Db.VM.get_resident_on ~__context ~self:vm with
+      | h when h <> Ref.null ->
+          Some h
+      | _ ->
+          None
+    in
+    host
+  in
+  match vms with
+  | [] | [_] ->
+      None
+  | vm :: _ ->
+      let host = get_host vm in
+      let same_host = List.for_all (fun vm' -> get_host vm' = host) vms in
+      debug "The breach state of group (UUID: %s) is :%b"
+        (Db.VM_group.get_uuid ~__context ~self:group)
+        same_host ;
+      if same_host then host else None
+
+let report_anti_affinity_alert ~__context ~group ~host =
+  let group_uuid = Db.VM_group.get_uuid ~__context ~self:group in
+  let host_uuid = Db.Host.get_uuid ~__context ~self:host in
+  let body =
+    Printf.sprintf
+      "<body><message>Breach on VM anti-affinity rules</message><VM_group>"
+    ^ group_uuid
+    ^ "</VM_group><host>"
+    ^ host_uuid
+    ^ "</host></body>"
+  in
+  let obj_uuid =
+    Db.Pool.get_uuid ~__context ~self:(Helpers.get_pool ~__context)
+  in
+  Xapi_alert.add ~msg:Api_messages.breach_on_vm_anti_affinity_rules ~cls:`Pool
+    ~obj_uuid ~body
+
+let get_all_records ~__context =
```

Comment:
Within the module `Xapi_vm_group_helpers`, this name is not good.
How about `anti_affinity_alerts`?